### PR TITLE
Store Orders: Display all taxes on orders, and add correct tax during refunds

### LINKED ISSUE #16634: Store Orders: Display collected tax on order details
It's missing presently:

screen shot 2017-07-27 at 10 06 03 am

Should probably also show it on a per line-item basis

### LINKED ISSUE #16645: Store Orders: Tax isn't included in refunds on line items
For orders with tax collected, you can't refund the tax:

screen shot 2017-07-27 at 11 57 58 am

See also #16634

### PR DESCRIPTION
**Pull request by ryelle**
Fixes #16645, fixes #16634

This PR has been pretty drastically updated. It used to be:

> This PR updates the price computed by changing the quantity in the refund modal to include tax on the item. It does check first that tax is not already included on the item (based on the API response, I couldn't actually get my test site into a state where it included tax on the products).

Now it's a refactor of the order details page, specifically the discount/shipping/total rows. It also includes these refund-with-tax fixes. The two big changes are:

- Now the discount, shipping, and total rows also have a Tax column, which line up with the product list table's tax and total columns.
- The refund modal, which also uses this table component, has a separate row for the refund input, so that users can see the shipping total and tax values.

Details table:
screen shot 2017-08-04 at 6 56 02 pm

Refund modal:
screen shot 2017-08-04 at 6 55 29 pm

To test:

- With a site that has taxes enabled, create an order or view an existing taxed order
- On the details view, make sure the discount, shipping, total, and refunds display as expected
- Try to give a refund on the order by changing the quantity in the refunds modal
- The price refunded should include the tax per-item
- You can see the shipping total and tax, and can enter that into the shipping refund input
- Refunding now, you can refund the entire order cost.

### CODE REVIEW

**Comment by allendav** [comment]
It seems to be over calculating the refund amount for me:

screen shot 2017-07-27 at 5 34 00 pm

screen shot 2017-07-27 at 5 34 31 pm

screen shot 2017-07-27 at 5 34 39 pm

**Comment by ryelle** [comment]
Looks like the item's `total_tax` is not per-item. I'll update this PR tomorrow to adjust for that.

**Comment by ryelle** [comment]
I've updated the PR to account for the tax bug, can you re-review @allendav ?

**Comment by allendav** [comment]
Yep - taking a look now

**Comment by allendav** [comment]
Something is still not right. My tax rate is 0.092

screen shot 2017-08-02 at 3 35 03 pm

If I attempt to refund the entire order by setting the quantity to 3 and the shipping to 5, I should get 333.06, but instead the refund view returns 332.60 which suggests the shipping tax (0.46) was not included

File: client/extensions/woocommerce/app/order/order-details-table.js [hunk 1]
```diff
@@ -80,7 +86,7 @@ class OrderDetailsTable extends Component {
 				<TableItem isHeader className="order__detail-item-product">{ translate( 'Product' ) }</TableItem>
 				<TableItem isHeader className="order__detail-item-cost">{ translate( 'Cost' ) }</TableItem>
 				<TableItem isHeader className="order__detail-item-quantity">{ translate( 'Quantity' ) }</TableItem>
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/order/order-details-table.js]: Hi! I've found a possible matching string that has already been translated **4** times:
`translate( 'Quantity:' )` *ES Score: 26*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/order/order-refund-row.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+	render() {
+		const { order, translate } = this.props;
+		const refundValue = order.refunds.length ? this.getRefundedTotal( order ) : false;
+		if ( ! refundValue ) {
+			return null;
+		}
+
+		return (
+			<div className="order__details-total-refund">
+				<div className="order__details-totals-label">{ translate( 'Refunded' ) }</div>
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/order/order-refund-row.js]: Hi! I've found a possible matching string that has already been translated **37** times:
`translate( 'Refunded To' )` *ES Score: 12*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/fix/store-16645-add-tax-to-refunds)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

File: client/extensions/woocommerce/app/order/order-shipping-row.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+	render() {
+		const { order, translate } = this.props;
+		if ( ! order ) {
+			return null;
+		}
+
+		return (
+			<div className="order__details-total-shipping">
+				<div className="order__details-totals-label">{ translate( 'Shipping' ) }</div>
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/order/order-shipping-row.js]: Hi! I've found a possible matching string that has already been translated **3** times:
`translate( 'Shipping address' )` *ES Score: 10*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by ryelle** [comment]
I've pushed up a refactor of the details table which displays more tax information and also tweaks the refund modal display. The description has been updated with more details and testing instructions.

**Comment by allendav** [comment]
Hey @ryelle - the final refund amount still seems to be missing the tax on the shipping?

screen shot 2017-08-07 at 12 24 30 pm

**Comment by kellychoffman** [comment]
I see the usefulness for a tax column when taxes are applied to discounts, orders, and shipping. When there isn't, the display output a bit overwhelming:

screen shot 2017-08-07 at 1 54 06 pm

When all taxes are set to $0.00, can we just not display them? So you would have: 

screen shot 2017-08-07 at 1 56 30 pm

**Comment by kellychoffman** [comment]
Question behind the reasoning for adding the Taxes column. Isn't it only useful when giving refunds? Is there a reason to have it displayed in the initial order view and not just in the refund modal?

**Comment by kellychoffman** [comment]
One more question. What does this input do? 

screen shot 2017-08-07 at 2 11 57 pm

I assume it is a way to allow a merchant to refund any amount? If so, I think we should add a label that indicates that, similar to how we do the 'Total refund amount' label.

**Comment by ryelle** [comment]
@allendav You're specifically saying only refund $5 for shipping, so it's only refunding the $5. If you put $5.46 you'll get the correct refund amount.

@kellychoffman: 

>  Is there a reason to have it displayed in the initial order view and not just in the refund modal?

I agree with hiding the col when it's all `0.00`, but I do think this info is useful for shop owners. For one thing, if you wanted to check that the "total total" is correct, you'd need that tax info - see Allen's shipping example - if we hide the tax col, the $0.46 shipping tax is a mystery addition to the total.

> I assume it is a way to allow a merchant to refund any amount?

Functionally yes, but it's meant to be the Shipping refund input, so the shop owner sees the shipping cost right above and can refund a partial shipping amount, or the entire shipping cost. While working on this I thought maybe a checkbox or button to "refund all shipping" might make more sense than this input, but I was trying to stick to the original design.

This is really hacked together, but something like:
screen shot 2017-08-08 at 11 12 17 am

On the other hand, now people wouldn't be able to give partial shipping refunds. But if the "Total refund amount" gets fixed so that it's editable, maybe that's not a problem.

**Comment by kellychoffman** [comment]
> I agree with hiding the col when it's all 0.00, but I do think this info is useful for shop owners. For one thing, if you wanted to check that the "total total" is correct, you'd need that tax info - see Allen's shipping example - if we hide the tax col, the $0.46 shipping tax is a mystery addition to the total.

Absolutely agree. My only concern was the column of zeros. If we are both in agreement on hiding that when there's no tax amount, all is good.

**Comment by jameskoster** [comment]
I agree the shipping refund input is confusing as it exists in this PR. How about we keep it inline like it is on live but pre-populate it with the total shipping cost? 

That way it's clear what the input is for whilst essentially doing the same job as a pre-checked checkbox. That'd provide the flexibility for partial shipping refunds too.

File: client/extensions/woocommerce/app/order/order-refund-row.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+				<div className="order__details-totals-tax"></div>
+				<div className="order__details-totals-value">
+					{ formatCurrency( refundValue, order.currency ) || refundValue }
```
Comment (justinshreve) [inline client/extensions/woocommerce/app/order/order-refund-row.js]: It's not a requirement for this PR, but could you log an issue to introduce a wrapper function that handles the `||` part of these calls? I know we do this in a few places and will probably have more as we expand Store, so making this simpler would be good.

Comment (ryelle) [inline client/extensions/woocommerce/app/order/order-refund-row.js]: Looks like as of a month ago we don't need to worry about it anymore – #15492 adds a default currency object so it defaults to a USD format if nothing's found.

Comment (justinshreve) [inline client/extensions/woocommerce/app/order/order-refund-row.js]: Oh perfect 👍

**Comment by ryelle** [comment]
@kellychoffman Should we also hide tax on the line items, if there are no taxes on the order? I don't know if totally removing a column on a per-order basis might be more confusing than it's worth to "clean up" the table.

**Comment by kellychoffman** [comment]
Do you mean removing this column when there is 0 tax? If so, no: I think we should keep the tax column even when there's 0 tax. 

screen shot 2017-08-09 at 11 55 23 am

**Comment by ryelle** [comment]
I've updated the branch to remove the tax column on the totals when they're all `0.00` (no tax is specified on this order). The tax col is left in the product line items table, though:

screen shot 2017-08-09 at 4 07 22 pm
screen shot 2017-08-09 at 4 05 17 pm
screen shot 2017-08-09 at 4 05 03 pm

This also updates the shipping refund input, pre-filling the input with total + tax.

File: client/extensions/woocommerce/app/order/order-shipping-row.js [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+		const tax = (
+			<div className="order__details-totals-tax">
+				{ formatCurrency( order.shipping_tax, order.currency ) }
+			</div>
+		);
+
+		return (
+			<div className="order__details-total-shipping">
+				<div className="order__details-totals-label">{ translate( 'Shipping' ) }</div>
```
Comment (a8ci18n) [inline client/extensions/woocommerce/app/order/order-shipping-row.js]: Hi! I've found a possible matching string that has already been translated **4** times:
`translate( 'Shipping address' )` *ES Score: 10*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

**Comment by allendav** [comment]
@ryelle wrote:

> @allendav You're specifically saying only refund $5 for shipping, so it's only refunding the $5. If you put $5.46 you'll get the correct refund amount.

I went to try that with the updated branch just now, but now the item tax is missing from the total

now-no-item-tax

As for entering the shipping + shipping tax - we don't display that sum anywhere - the merchant will have to do math to get that sum to enter it, and shipping will often not be a convenient whole number.

**Comment by allendav** [comment]
@jameskoster wrote

> I agree the shipping refund input is confusing as it exists in this PR. How about we keep it inline like it is on live but pre-populate it with the total shipping cost?

Oooh - I like this idea

**Comment by ryelle** [comment]
@allendav  are you sure you're up-to-date in this branch? If you have taxes, it should be displaying them in the totals area, and should be pre-populating the shipping box. Can you double check your API response, too? `{site.com}/wp-json/wc/v3/orders/{orderId}`, look for `tax_lines`

**Pull request merged**